File: src/main/java/persistence/sql/ddl/mapping/Table.java
package persistence.sql.ddl.mapping;

import jakarta.persistence.Entity;
import persistence.sql.ddl.exception.NotExistException;

public class Table {

    private final String name;

    public static Table from(Class<?> clazz) {
        if (!clazz.isAnnotationPresent(Entity.class)) {
            throw new NotExistException("@Entity annotation. class = " + clazz.getName());
        }
        
        if (clazz.isAnnotationPresent(jakarta.persistence.Table.class)) {
            jakarta.persistence.Table annotation = clazz.getAnnotation(jakarta.persistence.Table.class);
            return new Table(annotation.name());
        }

        Entity annotation = clazz.getAnnotation(Entity.class);
        if (!annotation.name().isEmpty()) {
            return new Table(annotation.name());
        }

        return new Table(clazz.getSimpleName());
    }

    public Table(String name) {
        this.name = name;
    }

    public String name() {
        return name;
    }
    
}

File: src/main/java/persistence/sql/ddl/Table.kt
package persistence.sql.ddl

import jakarta.persistence.Table

class Table (
    private val clazz: Class<*>
) {
    private val simpleName = clazz.simpleName.lowercase()
    private val columns: List<Column> = extractColumn(clazz)

    fun createQuery(): String {
        return "CREATE TABLE ${extractTableName()} (${columnsToQuery()})"
    }

    fun dropQuery(): String {
        return "DROP TABLE ${extractTableName()}"
    }

    private fun extractTableName(): String {
        return this.clazz.getAnnotation(Table::class.java)?.name ?: this.simpleName
    }

    private fun columnsToQuery(): String = columns.map { it.createQuery() }
        .filter { it.isNotEmpty() }
        .joinToString(",")

    private fun extractColumn(clazz: Class<*>): List<Column> =
        clazz.declaredFields.map { Column(it) }

}
